share/CMakeJS.cmake

```
@@ -0,0 +1,642 @@
+
+(This should wrap the CMakeLists.txt-side requirements for building a Napi Addon)
+]=============================================================================]#
+function(cmakejs_create_napi_addon name)
+
+    # Avoid duplicate target names
+    if(TARGET ${name})
+        message(SEND_ERROR "'cmakejs_create_napi_addon()' given target '${name}' which is already exists. Please choose a unique name for this Addon target.")
+        return()
+    endif()
+
+    set(options)
+    set(args ALIAS NAMESPACE NAPI_VERSION)
+    set(list_args)
+    cmake_parse_arguments(ARG "${options}" "${args}" "${list_args}" "${ARGN}")
+
+    # Generate the identifier for the resource library's namespace
+    set(ns_re "[a-zA-Z_][a-zA-Z0-9_]*")
+
+    if(NOT DEFINED ARG_NAMESPACE)
+        # Check that the library name is also a valid namespace
+        if(NOT name MATCHES "${ns_re}")
+            message(SEND_ERROR "Library name is not a valid namespace. Specify the NAMESPACE argument")
+            return()
+        endif()
+        set(ARG_NAMESPACE "${name}")
+    else()
+        if(NOT ARG_NAMESPACE MATCHES "${ns_re}")
+            message(SEND_ERROR "NAMESPACE for ${name} is not a valid C++ namespace identifier (${ARG_NAMESPACE})")
+            return()
+        endif()
+    endif()
+
+    # Needs more validation...
+    if(DEFINED ARG_NAPI_VERSION AND (ARG_NAPI_VERSION LESS_EQUAL 0))
+        message(SEND_ERROR "NAPI_VERSION for ${name} is not a valid Integer number (${ARG_NAPI_VERSION})")
+        return()
+    endif()
+
+    if(NOT DEFINED ARG_NAPI_VERSION)
+        if(NOT DEFINED NAPI_VERSION)
+            # default NAPI version to use if none specified
+            set(NAPI_VERSION 8)
+        endif()
+        set(ARG_NAPI_VERSION ${NAPI_VERSION})
+    endif()
+
+    if(ARG_ALIAS)
+        set(name_alt "${ARG_ALIAS}")
+    else()
+        set(name_alt "${ARG_NAMESPACE}")
+    endif()
+
+    if(VERBOSE)
+        message(STATUS "Configuring Napi Addon: ${name}")
+    endif()
+
+    # Begin a new Napi Addon target
+
+    add_library(${name} SHARED)
+    add_library(${name_alt}::${name} ALIAS ${name})
+
+    target_link_libraries(${name} PRIVATE cmake-js::cmake-js)
+
+    set_property(
+      TARGET ${name}
+      PROPERTY "${name}_IS_NAPI_ADDON_LIBRARY" TRUE
+    )
+
+    set_target_properties(${name}
+      PROPERTIES
+
+      LIBRARY_OUTPUT_NAME "${name}"
+      PREFIX ""
+      SUFFIX ".node"
+
+      ARCHIVE_OUTPUT_DIRECTORY "${CMAKEJS_BINARY_DIR}/lib"
+      LIBRARY_OUTPUT_DIRECTORY "${CMAKEJS_BINARY_DIR}/lib"
+      RUNTIME_OUTPUT_DIRECTORY "${CMAKEJS_BINARY_DIR}/bin"
+
+      # # Conventional C++-style debug settings might be useful to have...
+      # LIBRARY_OUTPUT_NAME_DEBUG "d${name}"
+      # ARCHIVE_OUTPUT_DIRECTORY_DEBUG "${CMAKEJS_BINARY_DIR}/lib/Debug"
+      # LIBRARY_OUTPUT_DIRECTORY_DEBUG "${CMAKEJS_BINARY_DIR}/lib/Debug"
+      # RUNTIME_OUTPUT_DIRECTORY_DEBUG "${CMAKEJS_BINARY_DIR}/bin/Debug"
+    )
+
```

```suggestion
    if (MSVC)
        # setup delayload
        target_link_options(${name} PRIVATE "/DELAYLOAD:NODE.EXE")
        target_link_libraries(${name} PRIVATE delayimp)

        if (CMAKE_SYSTEM_PROCESSOR MATCHES "(x86)|(X86)")
            target_link_options(${name} PUBLIC "/SAFESEH:NO")
        endif()
    endif()
```

Some needed stuff to make windows happy